docs/design/2020-05-08-standardize-error-codes-and-messages.md

```diff
@@ -0,0 +1,172 @@
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+'''
+workaround = '''Check the status, monitoring data and log of the TiKV server.'''
+```
+
+Benefit:
+- It's easy to write.
+- It handles multiple lines well.
+- It's easy to implement (parse toml and render markdown content)
+
+Tradeoff:
+- The markdown content can't be previewed on writing.
+
+#### metafile in markdown
+
+```md
+## Code: 8005
+### Error
+Write Conflict, txnStartTS is stale
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check whether `tidb_disable_txn_auto_retry` is set to `on`. If so, set it to `off`; if it is already `off`, increase the value of `tidb_retry_limit` until the error no longer occurs.
+
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+
+Benefit:
+- It's easy to write.
+- It handles multiple lines well.
+- The markdown content can be previewed on writing.
+
+Tradeoff:
+- It's hard to implement (see example below)
+- It's not self-consistent (see example below)
+
+Tradeoff Example:
+
+```md
+## Code: 8005
+### Error
+Write Conflict, txnStartTS is stale
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+```
+
+As the syntax above, the 8005 block is the message part of 8110 block, so we expect it's result is the same as this toml:
+
+```toml
+[error.8005]
+error = '''Write Conflict, txnStartTS is stale'''
+message = '''Transactions in TiDB encounter write conflicts.'''
+workaround = '''
+## Code: 9005
+### Error
+Region is unavailable
+### Message
+A certain Raft Group is not available, such as the number of replicas is not enough.
+This error usually occurs when the TiKV server is busy or the TiKV node is down.
+### Workaround
+Check the status, monitoring data and log of the TiKV server.
+'''
+```
+
+In this case, we must define the grammar and write a parser. Writing parser increase much complexity. What's worse
+is that I found no grammar that fits it.
+
+#### Summary
+
+Through the above discussion, we recommend the toml version of metafile. 
+
+In addition, the error codes should be adding only in case of conflict between versions.
```

Comment:
You mean `added`?

Reply:
I think it's `adding only` not `added only`. It means the code can be added, but can't be removed and reuse.

Reply:
is it better to use `append only`?

Reply:
@amyangfei Yes, `append only` is better